File: crates/api/src/tokens.rs
use axum::{
    Json, Router,
    extract::{Path, State},
    http::StatusCode,
    routing::post,
};
use mdp_auth::middleware::AuthUser;
use mdp_common::{error::AppError, response::ApiResponse};
use mdp_core::api_token::{ApiTokenCreated, ApiTokenInfo, ApiTokenService};
use serde::Deserialize;
use uuid::Uuid;

use crate::state::AppState;

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/tokens", post(create_token).get(list_tokens))
        .route("/tokens/{id}", axum::routing::delete(delete_token))
}

#[derive(Deserialize)]
struct CreateTokenRequest {
    name: String,
    permissions: Option<String>,
    expires_at: Option<chrono::DateTime<chrono::Utc>>,
}

async fn create_token(
    State(state): State<AppState>,
    auth: AuthUser,
    Json(req): Json<CreateTokenRequest>,
) -> Result<ApiResponse<ApiTokenCreated>, AppError> {
    let token = ApiTokenService::create(
        &state.db,
        auth.user_id,
        &req.name,
        req.permissions.as_deref(),
        req.expires_at,
    )
    .await?;
    Ok(ApiResponse::ok(token))
}

async fn list_tokens(
    State(state): State<AppState>,
    auth: AuthUser,
) -> Result<ApiResponse<Vec<ApiTokenInfo>>, AppError> {
    let tokens = ApiTokenService::list_by_user(&state.db, auth.user_id).await?;
    Ok(ApiResponse::ok(tokens))
}

async fn delete_token(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(id): Path<Uuid>,
) -> Result<StatusCode, AppError> {
    ApiTokenService::delete(&state.db, auth.user_id, id).await?;
    Ok(StatusCode::NO_CONTENT)
}

File: web/src/api/tokens.ts
import { request } from './client'

export interface ApiTokenInfo {
  id: string
  name: string
  permissions: string
  expires_at: string | null
  last_used_at: string | null
  created_at: string
}

export interface ApiTokenCreated extends ApiTokenInfo {
  token: string
}

export interface CreateTokenRequest {
  name: string
  permissions?: string
  expires_at?: string
}

export async function createToken(data: CreateTokenRequest): Promise<ApiTokenCreated> {
  return request<ApiTokenCreated>('/tokens', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  })
}

export async function listTokens(): Promise<ApiTokenInfo[]> {
  return request<ApiTokenInfo[]>('/tokens')
}

export async function deleteToken(id: string): Promise<void> {
  return request<void>(`/tokens/${id}`, { method: 'DELETE' })
}
